File: lua/pulpero/core/socket/methods.lua
local Methods = {}
local Runner = require('model_runner')
local Router = require('router.router')
local History = require('history.manager')
local ToolManager = require('managers.tool.manager')
local tools = require('tools')
local Parser = require('parser')
local uv = require('luv')

function Methods.new(logger, model_manager, setup)
    local self = setmetatable({}, { __index = Methods })
    self.logger = logger
    self.model_manager = model_manager
    self.router = nil
    self.history = nil
    self.setup = setup
    self.is_ready = false
    self.enable = true
    return self
end

function Methods.service_is_ready(self)
    self.logger:debug("checking if the service is ready")
    local status = self.model_manager:get_status_from_file()
    self.logger:debug("Model status ", { status })
    if status == "completed" then
        self.logger:debug("Service download status is completed")
        if self.enable then
            self.logger:debug("Service is enable")
            return true
        else
            self.logger:debug("The machine spirit is sleeping", { enable = self.enable })
            return false
        end
    else
        self.logger:debug("The machine spirit is not ready yet")
        return false
    end
end

function Methods.adapter(self, request)
    local response = {
        requestId = request.id,
        result = nil,
        error = nil
    }
    local method = request.method
    if method == "talk_with_model" then
        response = self:execute(function(methods)

            local file_context_data = {}
            self.logger:debug("talk with model request ", request)

            if request.params.file_context_data == nil then
                file_context_data = {
                    current_working_dir = "",
                    current_file_name = "",
                    current_file_path = ""
                }
            else
                file_context_data = request.params.file_context_data
            end

            return methods.router:route(request.params.message, file_context_data)
        end, response, method)
    elseif method == "prepear_env" then
        local function prepear_env(methods)
            if not methods.is_ready then
                local config = methods.setup:prepear_env()

                local tool_manager = ToolManager.new(methods.logger)

                tool_manager:register_tool(tools.create_create_file_tool(methods.logger))
                tool_manager:register_tool(tools.create_get_file_tool(methods.logger))
                tool_manager:register_tool(tools.create_find_file_tool(methods.logger))

                local parser = Parser.new(methods.logger)
                local runner = Runner.new(config, methods.logger, parser)
                methods.history = History.new(nil)

                methods.router = Router.new(config, methods.logger, runner, tool_manager, methods.history)
                methods.is_ready = true
            end
            return methods.is_ready
        end
        local success, result = pcall(prepear_env, self)
        if success then response.result = result else response.error = result end
        return response
    elseif method == "clear_model_cache" then
        response = self:execute(function(methods)
            methods.history:clear()
            return true
        end, response, method)
    elseif method == "get_download_status" then
        response = self:execute(function(methods)
            return methods.model_manager:get_download_status()
        end, response, method)
    elseif request.method == "get_service_status" then
        response = self:execute(function(methods)
            local status = methods.model_manager:get_status_from_file()
            return {
                running = true,
                model_ready = methods:service_is_ready(),
                download_status = status,
                pid = uv.os_getpid()
            }
        end, response, method)
    elseif request.method == "toggle" then
        self.enable = not self.enable
        response.result = self.enable
    else
        response.error = "Unknown method: " .. (request.method or "nil")
    end
    return response
end

function Methods.execute(self, callback, response, name)
    if self:service_is_ready() then
        local success, result = pcall(callback, self)
        if success then
            response.result = result
        else
            self.logger:error("Something went wrong when executing " .. name)
            self.logger:error("Error: ", result)
            response.result = false
            response.error = result
        end
    else
        self.logger:debug("service is not ready yet to " .. name)
        response.result = false
        response.error = "Service not ready - model still loading"
    end
    return response
end

return Methods

File: core/socket/methods.py
from core.managers.history.manager import HistoryManager
from core.managers.tool.manager import ToolManager
from core.runner.model.model_runner import Runner
from core.runner.model.parser import Parser
from core.util.logger import Logger
from core.managers.model import ModelManager
from core.socket.server import ServerRequest, ServerResponse
from core.router.router import FileContextData, RouterManager
from core.managers.tool.tools import tools

class Methods:

    logger: Logger
    model_manager: ModelManager
    router: RouterManager | None
    history: HistoryManager | None
    setup: SetupManager
    is_ready: bool
    enable: bool

    def __init__(self, logger: Logger, model_manager: ModelManager, setup) -> None:

        self.logger = logger
        self.model_manager = model_manager
        self.router = None
        self.history = None
        self.setup = setup
        self.is_ready = False
        self.enable = True

    def service_is_ready(self) -> bool:
        self.logger.info("checking if the service is ready")
        status = self.model_manager.get_status_from_file()
        self.logger.debug(f"Model status { status }")
        if status is "completed":
            self.logger.debug("Service download status is completed")
            if self.enable is True:
                self.logger.debug("Service is enable")
                return True
            else:
                self.logger.debug(f"The machine spirit is sleeping { self.enable }" )
                return False
        else:
            self.logger.debug("The machine spirit is not ready yet")
            return False

    def adapter(self, request: ServerRequest) -> ServerResponse:
        response = ServerResponse(requestId = request.Id, result = None, error = None)
        method = request.method

        match method:
            case "talk_with_model":
                if self.router is None:
                    response.error = 'router can not be None'
                    return response

                file_context: FileContextData = FileContextData(current_working_dir = '', current_file_name = '', current_file_path = '')
                if request.params.file_context is not None:
                    file_context = request.params.file_context
                response.result = self.router.route(request.params.message, file_context)
                return response

            case "get_live_code_feedback":
                if self.router is None:
                    response.error = 'router can not be None'
                    return response

                response.result = self.router.code_suggestion_pipeline(request.params.content, request.params.user_cursor)
                return response

            case "prepear_env":
                if not self.is_ready:
                    config = self.setup.prepear_env()

                    tool_manager = ToolManager(self.logger)

                    tool_manager.register_tool(tools['create_create_file_tool'](self.logger))
                    tool_manager.register_tool(tools['create_get_file_tool'](self.logger))
                    tool_manager.register_tool(tools['create_find_file_tool'](self.logger))

                    parser = Parser(self.logger)
                    runner = Runner(config, self.logger, parser)
                    self.history = HistoryManager(None)

                    self.router = RouterManager(self.logger, runner, tool_manager, self.history)
                    self.is_ready = True
                    response.result = self.is_ready
                return response

            case "clear_model_cache":
                if self.history is not None:
                    self.history.clear()
                response.result = True
                return response

            case "get_download_status":
                response.result = self.model_manager.get_download_status()
                return response

            case "get_service_status":
                status = self.model_manager.get_status_from_file()
                service_status = self.service_is_ready()
                response.result = f'{{ running: True, model_ready: {service_status}, download_status: {status} }}'
                return response

            case "toggle":
                self.enable = not self.enable
                response.result = self.enable
                return response

            case _:
                response.error = f"Unknown method: {request.method}"
                return response
